### PR TITLE
MODKBEKBJ-285 - Implement GET /loadHoldings/status

### PR DESCRIPTION
**Pull request by natali-zaitseva**
## Purpose
Due to https://issues.folio.org/browse/MODKBEKBJ-285 we want to have the ability to track loading status during holdings loading.

## Approach
* added new endpoint definition to descriptor and raml files
* added database trigger to have started date equal during all holdings loading stages
* added implementation form GET /loadHoldings/status endpoint
* added tests

### CODE REVIEW

File: src/main/java/org/folio/rest/impl/TenantApiImpl.java [hunk 1]
```diff
@@ -28,17 +38,44 @@
...
+    getVoidFuture(TenantTool.calculateTenantId(headers.get(OKAPI_TENANT_HEADER)))
+      .thenAccept(future::complete)
+      .exceptionally(e -> {
+        future.fail(e.getCause());
+        return null;
+      });
+    return future;
+  }
+
+  private CompletableFuture<Void> getVoidFuture(String tenantId) {
```
Comment (dmtkachenko) [inline src/main/java/org/folio/rest/impl/TenantApiImpl.java]: consider changing the name to something meaningful

File: src/main/java/org/folio/service/holdings/HoldingsServiceImpl.java [hunk 2]
```diff
@@ -38,21 +42,30 @@
...
+    this.holdingsStatusRepository = holdingsStatusRepository;
     this.delay = delay;
     this.retryCount = retryCount;
   }
 
-  public CompletableFuture<Void> loadHoldings(RMAPITemplateContext context, String tenantId) {
+  public CompletableFuture<Void> loadHoldings(RMAPITemplateContext context) {
+    final String tenantId = context.getOkapiData().getTenant();
     return populateHoldings(context)
+      .thenAccept(o -> updateStatus(getStatusPopulatingStagingArea(), tenantId))
```
Comment (dmtkachenko) [inline src/main/java/org/folio/service/holdings/HoldingsServiceImpl.java]: let's update the status inside populateHoldings method() to keep the consistency with loadHoldings()

File: src/main/java/org/folio/rest/impl/TenantApiImpl.java [hunk 3]
```diff
@@ -49,22 +50,32 @@ public TenantApiImpl() {
...
 
 
-  private void setLoadingStatus(Map<String, String> headers) {
-    String tenantId = TenantTool.calculateTenantId(headers.get(OKAPI_TENANT_HEADER));
-    holdingsStatusRepository.delete(tenantId);
-    holdingsStatusRepository.save(getStatusNotStarted(), tenantId);
+  private Future<Void> setLoadingStatus(Map<String, String> headers) {
+
+    Future<Void> future = Future.future();
+    getVoidFuture(TenantTool.calculateTenantId(headers.get(OKAPI_TENANT_HEADER)))
```
[the remark this answers is not in the document]
Comment (natali-zaitseva) [reply src/main/java/org/folio/rest/impl/TenantApiImpl.java]: updated, thanks @dmtkachenko

File: src/main/java/org/folio/repository/holdings/status/HoldingsLoadingStatusUtils.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+import org.folio.rest.jaxrs.model.HoldingsLoadingStatus;
+import org.folio.rest.jaxrs.model.JsonapiErrorResponse;
+import org.folio.rest.jaxrs.model.LoadStatusAttributes;
+import org.folio.rest.jaxrs.model.LoadStatusData;
+import org.folio.rest.jaxrs.model.LoadStatusInformation;
+import org.folio.rest.jaxrs.model.LoadStatusNameDetailEnum;
+import org.folio.rest.jaxrs.model.LoadStatusNameEnum;
+
+public class HoldingsLoadingStatusUtils {
```
[the remark this answers is not in the document]
Comment (natali-zaitseva) [reply src/main/java/org/folio/repository/holdings/status/HoldingsLoadingStatusUtils.java]: got it, renamed

File: ramls/types/loadHoldings/statusAttributes.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  "title": "Load Status Attributes Schema",
+  "description": "Load Status Attributes Schema",
+  "javaType": "org.folio.rest.jaxrs.model.LoadStatusAttributes",
+  "type": "object",
+  "additionalProperties": false,
+  "properties": {
+    "started": {
+      "type": "string",
+      "description": "Holdings loading started time",
+      "example": "1999-12-31 14:59:59"
```
[the remark this answers is not in the document]
Comment (natali-zaitseva) [reply ramls/types/loadHoldings/statusAttributes.json]: will use CURRENT_TIMESTAMP(2)  from https://www.postgresql.org/docs/9.1/functions-datetime.html

File: src/main/java/org/folio/repository/holdings/status/HoldingsStatusRepositoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+  @Override
+  public CompletableFuture<HoldingsLoadingStatus> get(String tenantId) {
+
+    final String query = String.format(GET_HOLDINGS_STATUS, getHoldingsStatusTableName(tenantId));
+    PostgresClient postgresClient = PostgresClient.getInstance(vertx, tenantId);
+    LOG.info("Select holdings loading status = " + query);
+    Future<ResultSet> future = Future.future();
+    postgresClient.select(query, future);
+    return mapResult(future, this::mapStatus);
```
Comment (carolegodfrey) [inline src/main/java/org/folio/repository/holdings/status/HoldingsStatusRepositoryImpl.java]: For my understanding -- Are we just keeping the latest status record in the table or do we have multiple entries (history) as the status gets updated

Comment (natali-zaitseva) [reply src/main/java/org/folio/repository/holdings/status/HoldingsStatusRepositoryImpl.java]: yes @carolegodfrey you are right, we keeping just a single record in the table which represents the latest status. After the module assigned to the tenant, we save the status and then just update it.

**Pull request merged**